crates/sui-json-rpc/src/read_api.rs

```diff
@@ -325,15 +328,33 @@ impl ReadApi {
                 })
                 .collect::<Vec<TransactionEventsDigest>>();
 
-            // fetch events from the DB
-            let events = self
-                .transaction_kv_store
-                .multi_get_events(&event_digests_list)
-                .await
-                .map_err(|e| {
-                    Error::UnexpectedError(format!("Failed to call multi_get_events for transactions {digests:?} with event digests {event_digests_list:?}: {e:?}"))
-                })?
-                .into_iter();
+            // fetch events from the DB with retry
+            let backoff = ExponentialBackoff {
+                max_elapsed_time: Some(Duration::from_secs(30)),
+                ..ExponentialBackoff::default()
+            };
+            let events = retry(backoff, || async {
+                match self
+                    .transaction_kv_store
+                    .multi_get_events(&event_digests_list)
+                    .await
+                {
+                    Ok(events) if !events.is_empty() => Ok(events),
```

Comment:
What happens if you try and fetch a transaction that genuinely has no events? It seems like that would just always take 30 seconds unless I'm misunderstanding something.

Reply:
ah you are right, I assumed that event list must have been checked empty or not before hitting kv store.

Reply:
hmm actually this wont work either, this is eventsDigest instead of eventDigest so empty events might still have a digest ..? looking

Reply:
I think using the events digest on the effects is the right idea -- i.e. you get the digest for all events that is part of the effects digest, and you retry until the response you get from the kv store is a set of events that matches that digest (i.e. it's not the digest of each individual event we care about, it's the overall digest).